File: 3710-design-an-array-statistics-tracker/design-an-array-statistics-tracker.py
from sortedcontainers import SortedList as bst

class StatisticsTracker:

    def __init__(self):
        self.a = deque()
        self.n = 0
        self.s = 0
        self.count = Counter()
        self.pq = []
        self.ss = bst()

    def addNumber(self, v) -> None:
        self.ss.add(v)
        self.count[v] += 1
        heappush(self.pq, (-self.count[v], v))
        self.s += v
        self.a.append(v)
        self.n += 1
        
    def removeFirstAddedNumber(self) -> None:
        v = self.a.popleft()
        self.ss.remove(v)
        self.count[v] -= 1
        self.s -= v
        self.n -= 1

    def getMean(self) -> int:
        return self.s // self.n

    def getMedian(self) -> int:
        return self.ss[self.n // 2]

    def getMode(self) -> int:
        while self.pq and -self.pq[0][0] != self.count[self.pq[0][1]]: heappop(self.pq)
        return self.pq[0][1]

File: 3710-design-an-array-statistics-tracker/design-an-array-statistics-tracker.cpp
class cmp {
public:
    bool operator() (const pair<int, int>& a, const pair<int, int>& b) const {
        if (a.first != b.first) {
            return a.first < b.first;
        } else {
            return a.second > b.second;
        }
    }
};

class StatisticsTracker {
private:
    long long sum;
    deque<int> nums;
    multiset<int, greater<>> upper;             // min heap (for bigger nums)
    multiset<int> lower;                        // max heap (for smaller nums)
    unordered_map<int, int> numFreq;            // {num, freq}
    set<pair<int, int>, cmp> freqNum;           // {freq, num}

    void balanceHeaps() {
        if (upper.size() == lower.size() - 1) { // if upper over erase
            upper.insert(*prev(lower.end()));
            lower.erase(prev(lower.end()));
        }
        if (lower.size() == upper.size() - 2) { // if lower over erase
            lower.insert(*prev(upper.end()));
            upper.erase(prev(upper.end()));
        }
    }

public:
    StatisticsTracker() : sum(0) {}
    
    void addNumber(int number) {
        // maintain sum and deque
        sum += number;
        nums.push_back(number);

        // maintain freq
        int prevFreq = numFreq[number]++;
        int currFreq = prevFreq + 1;
        if (prevFreq != 0) {
            freqNum.erase({prevFreq, number});
        }
        freqNum.insert({currFreq, number});

        // maintain two heap
        upper.insert(number);
        lower.insert(*prev(upper.end()));
        upper.erase(prev(upper.end()));
        if (upper.size() < lower.size()) {
            upper.insert(*prev(lower.end()));
            lower.erase(prev(lower.end()));
        }
    }
    
    void removeFirstAddedNumber() {
        // maintain sum and deque
        int number = nums.front(); // the number to remove
        nums.pop_front();
        sum -= number;
        
        // maintain freq
        int prevFreq = numFreq[number]--;
        int currFreq = prevFreq - 1;
        freqNum.erase({prevFreq, number});
        if (currFreq != 0) {
            freqNum.insert({currFreq, number});
        } else {
            numFreq.erase(number);
        }

        // maintain two heap
        if (number >= *prev(upper.end())) {
            upper.erase(upper.find(number));
        } else {
            lower.erase(lower.find(number));
        }
        balanceHeaps();
    }
    
    int getMean() {
        return sum / nums.size();
    }
    
    int getMedian() {
        return *prev(upper.end());
    }
    
    int getMode() {
        return prev(freqNum.end())->second;
    }
};